indigo_app/views/works.py

```diff
@@ -303,31 +303,93 @@ def get_top_contributors(self):
     template_name_suffix = '_commencements'
     tab = 'commencements'
 
     def get_context_data(self, **kwargs):
         context = super(WorkCommencementsView, self).get_context_data(**kwargs)
         context['provisions'] = provisions = self.work.all_commenceable_provisions()
-        context['uncommenced_provisions'] = self.work.all_uncommenced_provisions()
         context['commencements'] = commencements = self.work.commencements.all().reverse()
         context['has_all_provisions'] = any(c.all_provisions for c in commencements)
         context['has_main_commencement'] = any(c.main for c in commencements)
-        context['everything_commenced'] = context['has_all_provisions'] or (context['provisions'] and not context['uncommenced_provisions'])
+        context['uncommenced_provisions_count'], context['total_provisions_count'] = self.add_commencement_info(provisions, commencements)
+        context['everything_commenced'] = context['has_all_provisions'] or (context['provisions'] and not context['uncommenced_provisions_count'])
+
+        provision_set = {}
+
+        def add_to_set(p):
+            provision_set[p.id] = p
+            for c in p.children:
+                add_to_set(c)
+        for p in provisions:
+            add_to_set(p)
 
-        provision_set = {p.id: p for p in provisions}
         for commencement in commencements:
             # rich description of provisions
-            commencement.provision_items = [provision_set.get(p, p) for p in commencement.provisions]
+            commencement.commenced_provisions = self.get_commenced_provisions(commencement)
             # possible options
-            commencement.possible_provisions = self.get_possible_provisions(commencement, commencements, provisions)
+            commencement.possible_provisions = self.get_possible_provisions(commencement, commencements)
 
         return context
 
-    def get_possible_provisions(self, commencement, commencements, provisions):
+    def add_info(self, p, commenced_provisions, upc, tpc):
+        for c in p.children:
+            upc, tpc = self.add_info(c, commenced_provisions, upc, tpc)
+
+        p.commenced = p.id in commenced_provisions
+        p.uncommenced_descendants = (any(not c.commenced for c in p.children)) or any(c.uncommenced_descendants for c in p.children)
+
+        tpc += 1
+        if not p.commenced:
+            upc += 1
+
+        return upc, tpc
+
+    def add_commencement_info(self, provisions, commencements):
+        commenced_provisions = [p for c in commencements for p in c.provisions]
+        uncommenced_provisions_count = 0
+        total_provisions_count = 0
+
+        for p in provisions:
+            uncommenced_provisions_count, total_provisions_count = self.add_info(p, commenced_provisions, uncommenced_provisions_count, total_provisions_count)
+
+        return uncommenced_provisions_count, total_provisions_count
+
+    def get_commenced_provisions(self, commencement):
```

Comment:
This isn't actually getting all the commencements, it's getting all the commenceable provisions at this date, and marking those that are commenced by this commencement.

Since both this and `get_possible_provisions` actually operate on `all_commencable_provisions`, maybe call `all_commenceable_provisions` once and pass the provisions in. Then this (and the next fn) can be more accurately named `mark_commenced` and `mark_visible`.